### PR TITLE
feat: Delta table partition watermarks

### PR DESCRIPTION
**Pull request by harmw**
### Summary of Changes

Introducing watermark detection on Delta tables. Since [1] I lost track of this one around the time of going monorepo, here it is once more 🙈 

[1] https://github.com/amundsen-io/amundsendatabuilder/pull/427

### Tests

Testing for watermarks.

### Documentation

Nothing.

### CheckList
Make sure you have checked **all** steps below to ensure a timely review.
- [x] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
    - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [x] PR includes a summary of changes.
- [x] PR adds unit tests, updates existing unit tests, __OR__ documents why no test additions or modifications are needed.
- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public functions and the classes in the PR contain docstrings that explain what it does

### CODE REVIEW

File: requirements-dev.txt
Comment (mgorsk1) [inline requirements-dev.txt:16]: does it need to be a hard pin or will this work with `>=` ?

File: databuilder/databuilder/extractor/delta_lake_metadata_extractor.py
Comment (mgorsk1) [inline databuilder/databuilder/extractor/delta_lake_metadata_extractor.py:435]: I think it doesn't really matter which watermark is high and which low so this method could just return Optional[List[Watermark]] which would be more readable and simple. wdyt?

File: databuilder/databuilder/extractor/delta_lake_metadata_extractor.py
Comment (mgorsk1) [inline databuilder/databuilder/extractor/delta_lake_metadata_extractor.py:211]: in regards to comment on line 435 I would just do:
```
for watermark in watermarks:
    yield watermark
```

**Comment by mgorsk1** [comment]
by going through this code I realised that this is actually `SparkCatalogExtractor`, not particularly `DeltaExtractor`. Wdyt about generalizing it @harmw @feng-tao @samshuster

**Comment by feng-tao** [comment]
@mgorsk1 i don't think so given the extractor is for delta only which requires a databricks cluster to execute. I think we could have a separate extractor for standalone spark but let's not change the scope of this pr.

**Comment by mgorsk1** [comment]
I didn't mean it to be in the scope for this PR, there is just not much code here that makes it delta specific, just iterating over sparkCatalog databases and tables which might as well be for alternative hive metastore extraction.

**Comment by samshuster** [comment]
Yeah all good points. At the time that I wrote this extractor, I wasn't
aware of the python libraries that supported stand alone delta operations
without spark.
Something like this:
https://github.com/delta-io/delta-rs

I think ideally this extractor is rewritten to use that library which would
also enable getting rid of spark dependency and should also greatly speed
up this implementation.

Unfortunately as you know the hive metastore extractor will not work with
delta tables because most of the delta table metadata is not stored in
metastore but rather on cloud storage.

**Comment by harmw** [comment]
hm, cool, interesting 🤔 I simply picked up from where I left over a year ago, get the watermarks in and have it (Amundsen) be more valuable to how we're using it in our shop 😂 

Dropping the changes in this PR in favour of something new (extracting-without-spark) sounds pretty reasonable, _but_ I personally don't have time to do that in a rather timely fashion :( 

Would it make sense to get these changes in (after resolving the code review notes) and work on the break-out in a separate PR? And if yes, would that necessitate an RFC  of some sorts?

**Comment by feng-tao** [comment]
agree, we shouldn't block this pr instead should just file a github issue for enhancement.

btw, could you fix the lint:
```
flake8 .
./tests/unit/extractor/test_deltalake_extractor.py:62: [E501] line too long (125 > 120 characters)
./tests/unit/extractor/test_deltalake_extractor.py:366: [E501] line too long (128 > 120 characters)
./tests/unit/extractor/test_deltalake_extractor.py:392: [E501] line too long (127 > 120 characters)
./databuilder/extractor/delta_lake_metadata_extractor.py:435: [C901] 'DeltaLakeMetadataExtractor.create_table_watermarks' is too complex (12)
make: *** [lint] Error 1
Makefile:11: recipe for target 'lint' failed
```

I will take a look

File: requirements-dev.txt
Comment (feng-tao) [reply requirements-dev.txt:16]: I thought we had pyspark deps in setup.py?

**Review by feng-tao — APPROVED** [review]

**Review by feng-tao — APPROVED** [review]

**Comment by feng-tao** [comment]
```
flake8 .
mypy .
databuilder/extractor/delta_lake_metadata_extractor.py:213: error: Value of type "Watermark" is not indexable
databuilder/extractor/delta_lake_metadata_extractor.py:214: error: Value of type "Watermark" is not indexable
databuilder/extractor/delta_lake_metadata_extractor.py:503: error: Argument 2 to "filter" has incompatible type "Optional[List[ScrapedColumnMetadata]]"; expected "Iterable[ScrapedColumnMetadata]"
databuilder/extractor/delta_lake_metadata_extractor.py:530: error: Incompatible return value type (got "List[Tuple[Watermark, Watermark]]", expected "Optional[List[Watermark]]")
tests/unit/extractor/test_deltalake_extractor.py:369: error: Argument 1 to "create_table_watermarks" of "DeltaLakeMetadataExtractor" has incompatible type "Optional[ScrapedTableMetadata]"; expected "ScrapedTableMetadata"
tests/unit/extractor/test_deltalake_extractor.py:370: error: Argument 1 to "len" has incompatible type "Optional[List[Watermark]]"; expected "Sized"
tests/unit/extractor/test_deltalake_extractor.py:371: error: Value of type "Optional[List[Watermark]]" is not indexable
tests/unit/extractor/test_deltalake_extractor.py:371: error: Argument 1 to "len" has incompatible type "Union[Watermark, Any]"; expected "Sized"
tests/unit/extractor/test_deltalake_extractor.py:372: error: Value of type "Optional[List[Watermark]]" is not indexable
tests/unit/extractor/test_deltalake_extractor.py:372: error: Value of type "Union[Watermark, Any]" is not indexable
tests/unit/extractor/test_deltalake_extractor.py:396: error: Argument 1 to "create_table_watermarks" of "DeltaLakeMetadataExtractor" has incompatible type "Optional[ScrapedTableMetadata]"; expected "ScrapedTableMetadata"
tests/unit/extractor/test_deltalake_extractor.py:397: error: Argument 1 to "len" has incompatible type "Optional[List[Watermark]]"; expected "Sized"
tests/unit/extractor/test_deltalake_extractor.py:398: error: Value of type "Optional[List[Watermark]]" is not indexable
tests/unit/extractor/test_deltalake_extractor.py:398: error: Argument 1 to "len" has incompatible type "Union[Watermark, Any]"; expected "Sized"
tests/unit/extractor/test_deltalake_extractor.py:399: error: Value of type "Optional[List[Watermark]]" is not indexable
tests/unit/extractor/test_deltalake_extractor.py:399: error: Value of type "Union[Watermark, Any]" is not indexable
tests/unit/extractor/test_deltalake_extractor.py:439: error: Argument 1 to "create_table_watermarks" of "DeltaLakeMetadataExtractor" has incompatible type "Optional[ScrapedTableMetadata]"; expected "ScrapedTableMetadata"
Found 17 errors in 2 files (checked 361 source files)
```

**Comment by harmw** [comment]
@feng-tao should be all good now, finally 😅 

```First-time contributors need a maintainer to approve running workflows.```

Is this new?

**Comment by feng-tao** [comment]
@harmw , almost there :) :
```
 from collections import namedtuple
 from datetime import datetime
 from typing import (  # noqa: F401
-    Any, Dict, Iterator, List, Optional, Union, Tuple,
+    Any, Dict, Iterator, List, Optional, Tuple, Union,
 )
 
 from pyhocon import ConfigFactory, ConfigTree  # noqa: F401
Skipped 2 files
make: *** [isort_check] Error 1
```

**Comment by harmw** [comment]
completely ignore the contribution guidelines, my bad - should be good now, the steps in the `Makefile` succeeded. Future addition could be `make all` support, but whatever, all good now 🙂

**Comment by feng-tao** [comment]
trigger the CI button again, will merge once it is green

**Pull request merged**
